File: NextcloudTalk/Chat/NCMessageFileParameter.m
/**
 * SPDX-FileCopyrightText: 2020 Nextcloud GmbH and Nextcloud contributors
 * SPDX-License-Identifier: GPL-3.0-or-later
 */

#import "NCMessageFileParameter.h"

#import "NextcloudTalk-Swift.h"

@implementation NCMessageFileParameter

- (instancetype)initWithDictionary:(NSDictionary *)parameterDict
{
    self = [super initWithDictionary:parameterDict];
    if (self) {      
        self.path = [parameterDict objectForKey:@"path"];

        NSString *mimetype = [parameterDict objectForKey:@"mimetype"];

        if ([mimetype isKindOfClass:[NSString class]]) {
            self.mimetype = mimetype;
        }

        self.size = [[parameterDict objectForKey:@"size"] integerValue];
        self.previewAvailable = [[parameterDict objectForKey:@"preview-available"] boolValue];
        self.previewImageHeight = [[parameterDict objectForKey:@"preview-image-height"] intValue];
        self.previewImageWidth = [[parameterDict objectForKey:@"preview-image-width"] intValue];
        self.width = [[parameterDict objectForKey:@"width"] intValue];
        self.height = [[parameterDict objectForKey:@"height"] intValue];
        
        // NCChatFileStatus parameters
        NSString *fileId = [parameterDict objectForKey:@"fileId"];
        NSString *fileName = [parameterDict objectForKey:@"fileName"];
        NSString *filePath = [parameterDict objectForKey:@"filePath"];
        NSString *fileLocalPath = [parameterDict objectForKey:@"fileLocalPath"];

        if (fileId && fileName && filePath && fileLocalPath) {
            self.fileStatus = [[NCChatFileStatus alloc] initWithFileId:fileId fileName:fileName filePath:filePath fileLocalPath:fileLocalPath];
        } else {
            self.fileStatus = nil;
        }

        self.blurhash = [parameterDict objectForKey:@"blurhash"];
    }
    
    return self;
}

@end

File: NextcloudTalk/Chat/NCMessageFileParameter.swift
//
// SPDX-FileCopyrightText: 2026 Nextcloud GmbH and Nextcloud contributors
// SPDX-License-Identifier: GPL-3.0-or-later
//

@objcMembers
public class NCMessageFileParameter: NCMessageParameter {

    public var path: String?
    public var mimetype: String?
    public var size: Int?
    public var previewAvailable: Bool = false
    public var fileStatus: NCChatFileStatus?
    public var previewImageHeight: Int = 0
    public var previewImageWidth: Int = 0
    public var width: Int = 0
    public var height: Int = 0
    public var blurhash: String?

    override init?(dictionary dict: [String: Any]?) {
        super.init(dictionary: dict)

        self.path = dict?["path"] as? String
        self.mimetype = dict?["mimetype"] as? String
        self.previewAvailable = (dict?["preview-available"] as? String == "yes")

        self.size = dict?[intForKey: "size"] ?? 0
        self.previewImageHeight = dict?[intForKey: "preview-image-height"] ?? 0
        self.previewImageWidth = dict?[intForKey: "preview-image-width"] ?? 0
        self.width = dict?[intForKey: "width"] ?? 0
        self.height = dict?[intForKey: "height"] ?? 0

        self.blurhash = dict?["blurhash"] as? String

        if let fileId = dict?["fileId"] as? String, let fileName = dict?["fileName"] as? String,
           let filePath = dict?["filePath"] as? String, let fileLocalPath = dict?["fileLocalPath"] as? String {

            self.fileStatus = NCChatFileStatus(fileId: fileId, fileName: fileName, filePath: filePath, fileLocalPath: fileLocalPath)
        }
    }

}
